File: android-native/app/src/main/java/com/example/kpkn/data/repository/WikiLabRepository.kt
package com.example.kpkn.data.repository

import android.content.Context
import com.example.kpkn.data.db.*
import com.example.kpkn.data.prepopulateWikiLabAssets
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json

/**
 * WikiLab Repository — StateFlow + Room cache pattern.
 *
 * All static WikiLab data (muscles, joints, tendons, patterns, chains)
 * is loaded from Room on initialization and exposed as StateFlows.
 *
 * On first app launch, data is prepopulated from JSON assets.
 */
object WikiLabRepository {

    private lateinit var dao: WikiLabDao
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    // ─── StateFlow Caches ───────────────────────────────────────────────

    private val _muscles = MutableStateFlow<List<MuscleGroupEntity>>(emptyList())
    val muscles: StateFlow<List<MuscleGroupEntity>> = _muscles.asStateFlow()

    private val _joints = MutableStateFlow<List<JointEntity>>(emptyList())
    val joints: StateFlow<List<JointEntity>> = _joints.asStateFlow()

    private val _tendons = MutableStateFlow<List<TendonEntity>>(emptyList())
    val tendons: StateFlow<List<TendonEntity>> = _tendons.asStateFlow()

    private val _patterns = MutableStateFlow<List<MovementPatternEntity>>(emptyList())
    val patterns: StateFlow<List<MovementPatternEntity>> = _patterns.asStateFlow()

    private val _chains = MutableStateFlow<List<KineticChainEntity>>(emptyList())
    val chains: StateFlow<List<KineticChainEntity>> = _chains.asStateFlow()

    // ─── JSON Parser ────────────────────────────────────────────────────

    val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    // ─── Initialization ─────────────────────────────────────────────────

    fun initialize(context: Context, db: KpknDatabase) {
        dao = db.wikiLabDao()

        scope.launch {
            // Prepopulate on first run
            prepopulateWikiLabAssets(context, db)

            // Collect from Room → StateFlows
            launch { dao.getAllMuscles().collect { _muscles.value = it } }
            launch { dao.getAllJoints().collect { _joints.value = it } }
            launch { dao.getAllTendons().collect { _tendons.value = it } }
            launch { dao.getAllPatterns().collect { _patterns.value = it } }
            launch { dao.getAllChains().collect { _chains.value = it } }
        }
    }

    // ─── Lookup Helpers ─────────────────────────────────────────────────

    fun getMuscleById(id: String): MuscleGroupEntity? =
        _muscles.value.find { it.id == id }

    fun getJointById(id: String): JointEntity? =
        _joints.value.find { it.id == id }

    fun getTendonById(id: String): TendonEntity? =
        _tendons.value.find { it.id == id }

    fun getPatternById(id: String): MovementPatternEntity? =
        _patterns.value.find { it.id == id }

    fun getChainById(id: String): KineticChainEntity? =
        _chains.value.find { it.id == id }

    // ─── JSON Parsers ───────────────────────────────────────────────────

    fun parseStringList(jsonStr: String?): List<String> =
        try { json.decodeFromString<List<String>>(jsonStr ?: "[]") } catch (_: Exception) { emptyList() }

    fun parseInjuries(jsonStr: String?): List<InjuryInfo> =
        try { json.decodeFromString<List<InjuryInfo>>(jsonStr ?: "[]") } catch (_: Exception) { emptyList() }

    // ─── Filtered Queries ───────────────────────────────────────────────

    fun getMusclesByBodyPart(bodyPart: String): List<MuscleGroupEntity> =
        _muscles.value.filter { it.bodyPart == bodyPart }

    fun getJointsByBodyPart(bodyPart: String): List<JointEntity> =
        _joints.value.filter { it.bodyPart == bodyPart }

    fun getTendonsByMuscle(muscleId: String): List<TendonEntity> =
        _tendons.value.filter { it.muscleId == muscleId }

    fun getTendonsByJoint(jointId: String): List<TendonEntity> =
        _tendons.value.filter { it.jointId == jointId }

    // ─── Body Part Categories ──────────────────────────────────────────

    val bodyPartCategories: List<String> =
        listOf("upper", "lower", "core", "spine")

    fun getBodyPartLabel(bodyPart: String?): String = when (bodyPart) {
        "upper" -> "Tren Superior"
        "lower" -> "Tren Inferior"
        "core" -> "Core"
        "spine" -> "Columna"
        else -> "Otro"
    }

    fun getJointTypeLabel(type: String): String = when (type) {
        "hinge" -> "Bisagra"
        "ball-socket" -> "Esferoidea"
        "pivot" -> "Pivote"
        "gliding" -> "Deslizante"
        "saddle" -> "Silla de montar"
        "condyloid" -> "Condílea"
        else -> type
    }
}

// ─── Data Classes for Parsed JSON ──────────────────────────────────────────

@kotlinx.serialization.Serializable
data class InjuryInfo(
    val name: String,
    val description: String? = null,
    val riskExercises: List<String>? = null,
    val contraindications: List<String>? = null,
    val returnProgressions: List<String>? = null,
)

File: ios-native/KPKNFit/KPKNFit/Data/Repository/WikiLabRepository.swift
import Foundation

public final class WikiLabRepository {

    static let shared = WikiLabRepository()

    private var dao: WikiLabDao?
    private let db = KpknDatabase.instance()

    private var _muscles: [MuscleGroupEntity] = []
    private var _joints: [JointEntity] = []
    private var _tendons: [TendonEntity] = []
    private var _patterns: [MovementPatternEntity] = []
    private var _chains: [KineticChainEntity] = []

    private let musclesLock = NSLock()
    private let jointsLock = NSLock()
    private let tendonsLock = NSLock()
    private let patternsLock = NSLock()
    private let chainsLock = NSLock()

    public var muscles: [MuscleGroupEntity] {
        musclesLock.lock()
        defer { musclesLock.unlock() }
        return _muscles
    }

    public var joints: [JointEntity] {
        jointsLock.lock()
        defer { jointsLock.unlock() }
        return _joints
    }

    public var tendons: [TendonEntity] {
        tendonsLock.lock()
        defer { tendonsLock.unlock() }
        return _tendons
    }

    public var patterns: [MovementPatternEntity] {
        patternsLock.lock()
        defer { patternsLock.unlock() }
        return _patterns
    }

    public var chains: [KineticChainEntity] {
        chainsLock.lock()
        defer { chainsLock.unlock() }
        return _chains
    }

    private let jsonDecoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .useDefaultKeys
        return d
    }()

    private init() {}

    public static func getInstance() -> WikiLabRepository { shared }

    public func initialize() {
        let wikiLabDao = db.wikiLabDao
        self.dao = wikiLabDao

        Task { [weak self] in
            guard let self = self else { return }

            let muscles = await wikiLabDao.getAllMuscleGroups()
            let joints = await wikiLabDao.getAllJoints()
            let tendons = await wikiLabDao.getAllTendons()
            let patterns = await wikiLabDao.getAllMovementPatterns()
            let chains = await wikiLabDao.getAllKineticChains()

            self.musclesLock.lock()
            self._muscles = muscles
            self.musclesLock.unlock()

            self.jointsLock.lock()
            self._joints = joints
            self.jointsLock.unlock()

            self.tendonsLock.lock()
            self._tendons = tendons
            self.tendonsLock.unlock()

            self.patternsLock.lock()
            self._patterns = patterns
            self.patternsLock.unlock()

            self.chainsLock.lock()
            self._chains = chains
            self.chainsLock.unlock()
        }
    }

    public func getMuscleById(id: String) -> MuscleGroupEntity? {
        musclesLock.lock()
        defer { musclesLock.unlock() }
        return _muscles.first { $0.id == id }
    }

    public func getJointById(id: String) -> JointEntity? {
        jointsLock.lock()
        defer { jointsLock.unlock() }
        return _joints.first { $0.id == id }
    }

    public func getTendonById(id: String) -> TendonEntity? {
        tendonsLock.lock()
        defer { tendonsLock.unlock() }
        return _tendons.first { $0.id == id }
    }

    public func getPatternById(id: String) -> MovementPatternEntity? {
        patternsLock.lock()
        defer { patternsLock.unlock() }
        return _patterns.first { $0.id == id }
    }

    public func getChainById(id: String) -> KineticChainEntity? {
        chainsLock.lock()
        defer { chainsLock.unlock() }
        return _chains.first { $0.id == id }
    }

    public func parseStringList(jsonStr: String?) -> [String] {
        guard let jsonStr = jsonStr else { return [] }
        guard let data = jsonStr.data(using: .utf8) else { return [] }
        return (try? jsonDecoder.decode([String].self, from: data)) ?? []
    }

    public func parseInjuries(jsonStr: String?) -> [InjuryInfo] {
        guard let jsonStr = jsonStr else { return [] }
        guard let data = jsonStr.data(using: .utf8) else { return [] }
        return (try? jsonDecoder.decode([InjuryInfo].self, from: data)) ?? []
    }

    public func getMusclesByBodyPart(bodyPart: String) -> [MuscleGroupEntity] {
        musclesLock.lock()
        defer { musclesLock.unlock() }
        return _muscles.filter { $0.bodyPart == bodyPart }
    }

    public func getJointsByBodyPart(bodyPart: String) -> [JointEntity] {
        jointsLock.lock()
        defer { jointsLock.unlock() }
        return _joints.filter { $0.bodyPart == bodyPart }
    }

    public func getTendonsByMuscle(muscleId: String) -> [TendonEntity] {
        tendonsLock.lock()
        defer { tendonsLock.unlock() }
        return _tendons.filter { $0.muscleId == muscleId }
    }

    public func getTendonsByJoint(jointId: String) -> [TendonEntity] {
        tendonsLock.lock()
        defer { tendonsLock.unlock() }
        return _tendons.filter { $0.jointId == jointId }
    }

    public let bodyPartCategories: [String] = ["upper", "lower", "core", "spine"]

    public func getBodyPartLabel(bodyPart: String?) -> String {
        guard let bodyPart = bodyPart else { return "Otro" }
        switch bodyPart {
        case "upper": return "Tren Superior"
        case "lower": return "Tren Inferior"
        case "core": return "Core"
        case "spine": return "Columna"
        default: return "Otro"
        }
    }

    public func getJointTypeLabel(type: String) -> String {
        switch type {
        case "hinge": return "Bisagra"
        case "ball-socket": return "Esferoidea"
        case "pivot": return "Pivote"
        case "gliding": return "Deslizante"
        case "saddle": return "Silla de montar"
        case "condyloid": return "Condílea"
        default: return type
        }
    }
}

public struct InjuryInfo: Codable {
    public let name: String
    public let description: String?
    public let contraindications: [String]?
    public let returnProgressions: [String]?
    
    enum CodingKeys: String, CodingKey {
        case name, description, contraindications
        case returnProgressions = "returnProgressions"
    }
}
